src/AI/Muriel.java

```diff
@@ -524,88 +515,6 @@ private void queueUnitProductionActions(GameMap gameMap)
-      if( !enemyModel.hasDirectFireWeapon() )
-      {
-        // We currently ignore indirect-fire units for combat estimates.
-        enemyModels.remove(enemyModel);
-        enemyQuantity.remove(enemyModel);
-        continue;
-      }
-      Collections.sort(myModels, new UnitMatchupComparator(enemyModel, myUnitEffectMap, UnitMatchupComparator.ComparisonType.DAMAGE_RATIO) );
-
-      // Grab the least effective model we have vs enemyModel and figure out how much
-      // damage we can do with the numbers we have, assuming only the enemyModel and myModel
-      // numbers as participants.
-      UnitModel myModel = myModels.get(0);
-      if( !myModel.hasDirectFireWeapon() )
-      {
-        // We currently ignore indirect-fire units for combat estimates.
-        myQuantity.remove(enemyModel);
-        continue;
-      }
-      UnitMatchupAndMetaInfo umami = myUnitEffectMap.get(new UnitModelPair(myModel, enemyModel));
-      double enemyQuant = enemyQuantity.get(enemyModel);      // How many baddies?
-      double myQuant = myQuantity.get(myModel);               // How many goodies?
-      double myEffectiveQuant = myQuant * umami.damageRatio;  // How many of their guys do my guys equal?
-      double diffResult = enemyQuant - myEffectiveQuant;      // How much did I knock their numbers down?
-      double enemyLeft = (diffResult < 0) ? 0 : diffResult;   // How many of them are left?
-      double myEffectiveLeft = (diffResult < 0) ? -diffResult : 0;
-      double myLeft = myEffectiveLeft*(1.0/umami.damageRatio); // How many of mine?
-
-      log("Forces:");
-      log("  Mine");
-      log(myQuantity);
-      log("  Enemy");
-      log(enemyQuantity);
-      log(String.format("  Pitting my %sx%s against enemy %sx%s", myModel, myQuant, enemyModel, enemyQuant));
-      log(String.format("  %s of my %s are worth %s of their %s", myQuant, myModel, myEffectiveQuant, enemyModel));
-      log(String.format("  Their numbers are knocked down to %s", diffResult));
-      log(String.format("  Remains: I have %s and enemy has %s", myLeft, enemyLeft));
-
-      // Update stored amounts based on results.
-      if( 0 == enemyLeft )
-      {
-        log(String.format("Removing enemy model %s", enemyModel));
-        enemyQuantity.remove(enemyModel); // That enemy type is completely accounted for.
-        enemyModels.remove(enemyModel);
-      }
-      else
-      {
-        enemyQuantity.put(enemyModel, enemyLeft); // Didn't account for all enemies of this type; store the remainder.
-      }
-      if( 0 == myLeft )
-      {
-        log(String.format("Removing my %s", myModel));
-        myQuantity.remove(myModel); // Ran out of units of this type; hope we have some other unit model to fill the gap.
-        myModels.remove(myModel);
-      }
-      else
-      {
-        myQuantity.put(myModel, myLeft); // Didn't need all of my guys of this type to counter their guys; bring on whatever's next.
-      }
-    }
-  }
```

Comment:
I was using it to identify the "biggest threats" in the enemy army, but it ended up not being useful. If the enemy had 20 infantry, it would decide that one Md Tank was enough to deal with the threat, and not recommend building anything else. Chasing down 20 infantry with a single Md Tank is not fun.
The current approach tries to build something each turn to counter the most numerous enemy unit type, which fixes the "one unit is enough" philosophy, but I'm still not completely satisfied. It mostly works out because the most cost-efficient units against infantry (Md Tanks, followed by Neos and then AAs or Tanks I think) tend to be generally good against other things too. I may update this to work a different way at some point, time permitting.
But to your other point, yes, this could totally go into AIUtils, and would definitely be useful for "can I take that hill" type scenarios. However, the initial implementation was kinda hardwired into Muriel, and I didn't care to extract it at the time.